% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/seurat.R
\name{doHeatMap}
\alias{doHeatMap}
\title{Gene expression heatmap}
\usage{
doHeatMap(object, cells.use = NULL, genes.use = NULL, disp.min = -2.5,
  disp.max = 2.5, draw.line = TRUE, do.return = FALSE,
  order.by.ident = TRUE, col.use = pyCols, slim.col.label = FALSE,
  group.by = NULL, remove.key = FALSE, cex.col = NULL, do.scale = TRUE,
  ...)
}
\arguments{
\item{object}{Seurat object}

\item{cells.use}{Cells to include in the heatmap (default is all cells)}

\item{genes.use}{Genes to include in the heatmap (ordered)}

\item{disp.min}{Minimum display value (all values below are clipped)}

\item{disp.max}{Maximum display value (all values above are clipped)}

\item{draw.line}{Draw vertical lines delineating cells in different identity
classes.}

\item{do.return}{Default is FALSE. If TRUE, return a matrix of scaled values
which would be passed to heatmap.2}

\item{order.by.ident}{Order cells in the heatmap by identity class (default
is TRUE). If FALSE, cells are ordered based on their order in cells.use}

\item{col.use}{Color palette to use}

\item{slim.col.label}{if (order.by.ident==TRUE) then instead of displaying 
every cell name on the heatmap, display only the identity class name once 
for each group}

\item{group.by}{If (order.by.ident==TRUE) default,  you can group cells in 
different ways (for example, orig.ident)}

\item{remove.key}{Removes the color key from the plot.}

\item{\dots}{Additional parameters to heatmap.2. Common examples are cexRow
and cexCol, which set row and column text sizes}
}
\value{
If do.return==TRUE, a matrix of scaled values which would be passed
to heatmap.2. Otherwise, no return value, only a graphical output
}
\description{
Draws a heatmap of single cell gene expression using the heatmap.2 function.
}
