% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/seurat.R
\name{find.markers}
\alias{find.markers}
\title{Gene expression markers of identity classes}
\usage{
find.markers(object, ident.1, ident.2 = NULL, genes.use = NULL,
  thresh.use = log(2), test.use = "bimod", min.pct = 0,
  print.bar = TRUE, only.pos = FALSE)
}
\arguments{
\item{object}{Seurat object}

\item{ident.1}{Identity class to define markers for}

\item{ident.2}{A second identity class for comparison. If NULL (default) -
use all other cells for comparison.}

\item{genes.use}{Genes to test. Default is to use all genes.}

\item{thresh.use}{Limit testing to genes which show, on average, at least
X-fold difference (log-scale) between the two groups of cells.

Increasing thresh.use speeds up the function, but can miss weaker signals.}

\item{test.use}{Denotes which test to use. Seurat currently implements
"bimod" (likelihood-ratio test for single cell gene expression, McDavid et
al., Bioinformatics, 2011, default), "roc" (standard AUC classifier), "t"
(Students t-test), and "tobit" (Tobit-test for differential gene expression,
as in Trapnell et al., Nature Biotech, 2014)}

\item{min.pct}{- only test genes that are detected in a minimum fraction of min.pct cells 
in either of the two populations. Meant to speed up the function by not testing genes that are very infrequently expression}

\item{print.bar}{Print a progress bar once expression testing begins (uses pbapply to do this)}

\item{only.pos}{Only return positive markers (FALSE by default)}
}
\value{
Matrix containing a ranked list of putative markers, and associated statistics (p-values, ROC score, etc.)
}
\description{
Finds markers (differentially expressed genes) for identity classes
}
