% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/seurat.R
\name{genePlot}
\alias{genePlot}
\title{Scatter plot of single cell data}
\usage{
genePlot(object, gene1, gene2, cell.ids = NULL, col.use = NULL,
  pch.use = 16, cex.use = 1.5, use.imputed = FALSE, do.ident = FALSE,
  do.spline = FALSE, spline.span = 0.75, ...)
}
\arguments{
\item{object}{Seurat object}

\item{gene1}{First feature to plot. Typically gene expression but can also
be metrics, PC scores, etc. - anything that can be retreived with fetch.data}

\item{gene2}{Second feature to plot.}

\item{cell.ids}{Cells to include on the scatter plot.}

\item{col.use}{Colors to use for identity class plotting.}

\item{pch.use}{Pch argument for plotting}

\item{cex.use}{Cex argument for plotting}

\item{use.imputed}{Use imputed values for gene expression (Default is FALSE)}

\item{do.ident}{False by default. If TRUE,}

\item{do.spline}{Add a spline (currently hardwired to df=4, to be improved)}

\item{\dots}{Additional arguments to be passed to plot.}
}
\value{
No return, only graphical output
}
\description{
Creates a scatter plot of two features (typically gene expression), across a
set of single cells. Cells are colored by their identity class.
}
